### ISSUE TITLE
Mouse clicks "bleeds through" to toolbar of iconified window

### ISSUE THREAD

**Issue by GokMasE** [freemint/freemint#300]
If you iconify a window that has a toolbar attached, XaAES will still allow mouse clicks to the original windows toolbar - that is, in the area where the toolbar would have been, if the iconified window was supposed to have it active/visible.

When mouse clicks are directed to this "invisible toolbar", XaAES handles redrawing of the buttons (selected state on/off) which makes the buttons visibly appear as they are being clicked on. Example in screenshot below:

![toolbar_icon_bleed](https://user-images.githubusercontent.com/65512522/229375895-e050284a-4dbb-4231-b5e7-e1c9ac92104a.png)
_(In the current example, it seems like the "invisible toolbar" is located just below an "imaginary infobar")_

I imagine filtering out such inputs to the toolbar when window is iconified should be very trivial.

**Note: This bug partly exist in xaaes-ozk too (#241) - mouse input is allowed just the same as with xaaes-cur, but the toolbar buttons are not being redrawed as they are clicked.**

**Comment by xdelatour** [comment]
Thanks for your bug report.

> I imagine filtering out such inputs to the toolbar when window is iconified should be very trivial

I guess hiding the root object should be enough,
I will look next week

**Comment by xdelatour** [comment]
Here is a patch to hide/unhide the toolbar when `wind_set(WF_ICONIFY/WF_UNICONIFY)` is called.

[toolbar-v2.patch.txt](https://github.com/freemint/freemint/files/11181271/toolbar-v2.patch.txt)

I tried others methods without success (e.g. clear the `NO_TOPPED` flag to send `WM_TOPPED` instead of `MU_BUTTON`).
- I chose to not add these changes in `xa_wind.c` because `wind_set()` calls `iconify_window()/uniconify_window()` which are defined in `c_window.c`
- `uniconify_window()` already includes two statements in order to reinstall the menu and to [fix the toolbar](https://github.com/freemint/freemint/blob/af27cfd10deaf7c8043f02f655a9d0b749debb7d/xaaes/Changes.hk#L446-L448)

The only canveats is a potential bug if the toolbar is hidden for some reason BEFORE iconification. The toolbar will be drawn (unhide) after uniconification. Maybe I can improve the patch to save the old status of `OF_HIDETREE`?

Anyway, the toolbar management is...at least incomplete:
- `remove_window_widgets()` is unable to manage menu/toolbar [here](https://github.com/freemint/freemint/blob/af27cfd10deaf7c8043f02f655a9d0b749debb7d/xaaes/src.km/c_window.c#L1543) and [here](https://github.com/freemint/freemint/blob/af27cfd10deaf7c8043f02f655a9d0b749debb7d/xaaes/src.km/c_window.c#L1554)
- Called 'non-standard' [here](https://github.com/freemint/freemint/blob/af27cfd10deaf7c8043f02f655a9d0b749debb7d/xaaes/src.km/c_window.c#L1506-L1511) (as opposed to standard widgets)
- Most of the toolbar functions are in `win_draw.c`, `widgets.c`, `c_window.c`. These files ARE the window renderer. Nevertheless, the toolbar is basically an AES object tree. All other AES object are drawn by the AES object renderer (`render_obj.c`, `draw_obj.c`, `scrlobjc.c`, `obtree.c`). 

I wonder what the authors had planned to do? No doubt compatibility with N.AES 2.0, but why separate widgets and AES objects? And include features of N.AES 3.0 in X-Window before its release? Was the goal to merge GEM and X? In the same way that MiNT was moving towards full compatibility with Unix (as evidenced by already reserved SysV IPC system calls)

**Comment by GokMasE** [comment]
If someone could wrap up a test binary I'd be happy to check out the patch. @mikrosk?

Restoring the uniconified window to a state where the toolbar is hidden (if it was, prior to iconfication) seems like a good choice if possible. Suggested solution sounds like a good bet.

**Comment by th-otto** [comment]
Wouldn't it be easier to just ignore the mouse-clicks if the window is iconified? So you don't have to mess with the hidden state.

**Comment by xdelatour** [comment]
AFAIK all AES always send an event if the user clicks inside the work area. Moreover, an iconified window is the same window whose XaAES changed both size and attribute.

**Comment by xdelatour** [comment]
> Edit: currently digging through the sources, and i wonder if the same would happen if the window has a menubar installed?

Already tested a couple of days ago. No problem with the menubar

**Comment by mikrosk** [comment]
Latest xaaes with the patch above: [xaaes.km.zip](https://github.com/freemint/freemint/files/11189393/xaaes.km.zip)

**Comment by xdelatour** [comment]
Is it usual to install a toolbar and then hide it?

The management of the window menubar is very different from the toolbar, it seems impossible to manage their iconification in the same way.

I think the toolbar remains active because it belongs to the workarea. On the contrary, the menubar have its own row in the window structure (row3, below titlebar and infobar), which is removed when the widgets are updated (it's just a therory).

**Comment by th-otto** [comment]
I'm currently trying to hack together a small sample program to test this. I've never used the MENUBAR/WF_MENU features before, so i wonder: if i install a menubar with it, and select and entry, how do i get the title into its normal state again?
menu_tnormal() only seems to work for menu trees installed with menu_bar(), but not with wind_set(WF_MENU). And the comment at https://github.com/freemint/freemint/blob/master/xaaes/src.km/menuwidg.c#L1931 also indicates that XaAES does not do that automatically.

**Comment by GokMasE** [comment]
> And another question @GokMasE : how do i get a separate iconify window? With taskbar installed, the entry for the window in the taskbar is written in italics like this
> 
> ![Screenshot_20230410_142349](https://user-images.githubusercontent.com/12425871/230901048-f234be4f-29f1-4054-ad42-2bb33a2037d6.png)
> 
> But i don't get a separate iconify window.

Sorry, I seem to have missed this question back in April.
Seeing that Taskbar represents the application in italics style text indicates that the application window is hidden rather than iconified. Are you sure the window in question was sent wind_set(handle&,WF_ICONIFY,...)?

**Comment by skaftetryne** [comment]
Italics button title in Taskbar means that the *application* is hidden. Taskbar does not know of/care about individual windows.On 14 Aug 2023 19:22, Joakim ***@***.***> wrote:

And another question @GokMasE : how do i get a separate iconify window? With taskbar installed, the entry for the window in the taskbar is written in italics like this

But i don't get a separate iconify window.

Sorry, I seem to have missed this question back in April.
Seeing that Taskbar represents the application in italics style text indicates that the application window is hidden rather than iconified. Are you sure the window in question was sent wind_set(handle&,WF_ICONIFY,...)?

**Comment by GokMasE** [comment]
That is true, but as a consequence of hiding an app, all of its (currently open) windows are moved outside the visible screen. So normally a hidden app indicates that its windows are hidden.

Sure, there are examples of scenarios where you can open a new window for an app in hidden state (who then already have windows hidden/located off screen), most notably if the hidden application has a menu bar that stays active after the windows are hidden - then the user can often trigger a new window to open through a CTRL+N or similar. IIRC, such an app is considered still in hidden state, until all of its windows are located within the visible screen again.

But by far the most common situation is that hidden app equals hidden windows.

**Comment by mikrosk** [comment]
@GokMasE did the binary published in https://github.com/freemint/freemint/issues/300#issuecomment-1501594976 work for you? If so, perhaps we can merge @xdelatour's patch. If I feel lucky, maybe I can extend it to store/restore the OF_HIDETREE state, too (having a test app would certainly help).

**Comment by GokMasE** [comment]
I apologize for the the late reply, I finally got around to check up on this:

Yes, the binary mentioned above does work, and at first glance it does seem to sort the problem - but...

**It turns out that any double clicks sent to the iconified window when the mouse is located where the toolbar would have been placed (relative to the window title/mover) will be eaten, presumably by an "invisible toolbar".**

Double clicking in this exact area will not result in the iconified window to be un-iconified.
Move the mouse slightly above or slightly below, and the double click works normally again.

**Comment by mikrosk** [comment]
Hmm, that doesn't sound like ready to  merge. :-/ @xdelatour any chance for having a second look?

**Comment by mikrosk** [comment]
@GokMasE can you please test this build? it contains a different patch than the above: [xaaes.km.zip](https://github.com/user-attachments/files/28040177/xaaes.km.zip)
